pkg/runtime/cache/account_test.go

```diff
@@ -0,0 +1,145 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package cache_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zapcore"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	k8stesting "k8s.io/client-go/testing"
+	ctrlrtzap "sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	ackrtcache "github.com/aws/aws-controllers-k8s/pkg/runtime/cache"
+)
+
+const (
+	testNamespace = "ack-system"
+
+	testAccount1    = "012345678912"
+	testAccountARN1 = "arn:aws:iam::012345678912:role/S3Access"
+	testAccount2    = "219876543210"
+	testAccountARN2 = "arn:aws:iam::012345678912:role/root"
+)
+
+func TestAccountCache(t *testing.T) {
+	accountsMap1 := map[string]string{
+		testAccount1: testAccountARN1,
+	}
+
+	accountsMap2 := map[string]string{
+		testAccount1: testAccountARN1,
+		testAccount2: testAccountARN2,
+	}
+
+	// create a fake k8s client and a fake watcher
+	k8sClient := k8sfake.NewSimpleClientset()
+	watcher := watch.NewFake()
+	k8sClient.PrependWatchReactor("configMaps", k8stesting.DefaultWatchReactor(watcher, nil))
+
+	zapOptions := ctrlrtzap.Options{
+		Development: true,
+		Level:       zapcore.InfoLevel,
+	}
+	fakeLogger := ctrlrtzap.New(ctrlrtzap.UseFlagOptions(&zapOptions))
+
+	// initlizing account cache
+	accountCache := ackrtcache.NewAccountCache(k8sClient, fakeLogger)
+	stopCh := make(chan struct{})
+
+	// This is mainly used to avoid using time.Sleep
+	accountCache.Run(stopCh)
+
+	// Test create events
+	k8sClient.CoreV1().ConfigMaps(testNamespace).Create(
+		context.Background(),
+		&corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "random-map",
+			},
+			Data: accountsMap1,
+		},
+		metav1.CreateOptions{},
+	)
+
+	time.Sleep(time.Second)
```

Comment:
I'm not really comfortable using `time.Sleep()` calls in this test, and `SharedInformer.HasSynced` isn't really useful since it will always return `true` after the first informer sync.

Multiple other solutions are possible but all of them imply changes the Cache public functions signatures:

1. expose `SharedInformer.LastSyncResourceVersion()` and use something like
```go
lastKnownVersion := informer.LastSyncResourceVersion()
for informer.LastSyncResourceVersion() == lastKnownVersion {}
// informer has synced at this point
```

2. introduce a new parameter in `AccountCache.Run` that will take a `func()` type and will execute the function at each call
```go
func (c *AccountCache) Run(stopCh <-chan struct{}, fn func()) {
	c.informer.AddEventHandler(k8scache.ResourceEventHandlerFuncs{
		AddFunc: func(obj interface{}) {
			defer fn()
			if resourceMatchACKRoleAccountsConfigMap(obj) {
				//...
			}
		},
		// ....
	})
	go c.informer.Run(stopCh)
}
```

In the test file we can use it like this:
```go
accountCache := ackrtcache.NewAccountCache(k8sClient, fakeLogger)
stopCh := make(chan struct{})
eventCh := make(chan struct{})
accountCache.Run(stopCh, func() { eventCh <- struct{}{} })
k8sClient.CoreV1().ConfigMaps(testNamespace).Create(...)
<- eventCh // instead of time.Sleep(time.Second)
```

*NOTE*: i don't feel that this is the right approach

3. introduce a new parameter that will take am optional list of handlers and append them to the shared informer ResourceEventHandlers
```go
func (c *NamespaceCache) Run(stopCh <-chan struct{}, handlers ...k8scache.ResourceEventHandlerFuncs) {
	c.informer.AddEventHandler(k8scache.ResourceEventHandlerFuncs{
		// default handler
	})
	for _, handler := range handlers {
		c.informer.AddEventHandler(handler)
	}
	go c.informer.Run(stopCh)
}
```

using the same testing method in the previous option, we can append a new `EventResourceHandlerFunc` that will send to `eventCh` a `struct{}{}` value whenever the handlers are called.

Finally, IMHO i believe that putting these tests on the same package name `cache` will simplify a things a lot, at least we won't worry about making changes in the public functions

Reply:
I'm also starting to think if we really need to use `ResourceEventHandlerFuncs`... `SharedInformer` exposes a `GetStore` function that return a ["thread safe" store](https://github.com/kubernetes/client-go/blob/master/tools/cache/store.go#L132-L140) containing all the resources/methods we'll need to use. 
But anyways testing the cache without `time.Sleep` will require using the `SharedInformer.AddEventHandler` if we go for option .2 or .3

Reply:
We can certainly improve this in the future and remove the sleep calls, but this is perfectly fine for now. Nice work on this, @A-Hilaly